File: frontend/src/types/liability.ts
export type LiabilityType =
  | 'property_loan'
  | 'business_loan'
  | 'owner_loan'
  | 'family_loan'
  | 'other_liability';

export type LiabilitySourceType =
  | 'manual'
  | 'document_confirmed'
  | 'document_auto_created'
  | 'system_migrated';

export type LiabilityReportCategory =
  | 'darlehen_und_kredite'
  | 'sonstige_verbindlichkeiten';

export interface LiabilityRelatedTransaction {
  id: number;
  type: string;
  amount: number;
  transaction_date: string;
  description?: string;
}

export interface LiabilityRelatedRecurring {
  id: number;
  recurring_type: string;
  description: string;
  amount: number;
  frequency: string;
  is_active: boolean;
  next_generation_date?: string | null;
}

export interface LiabilityRecord {
  id: number;
  user_id: number;
  liability_type: LiabilityType;
  source_type: LiabilitySourceType;
  display_name: string;
  currency: string;
  lender_name: string;
  principal_amount: number;
  outstanding_balance: number;
  interest_rate?: number | null;
  start_date: string;
  end_date?: string | null;
  monthly_payment?: number | null;
  tax_relevant: boolean;
  tax_relevance_reason?: string | null;
  report_category: LiabilityReportCategory;
  linked_property_id?: string | null;
  linked_loan_id?: number | null;
  source_document_id?: number | null;
  is_active: boolean;
  can_edit_directly: boolean;
  can_deactivate_directly: boolean;
  edit_via_document: boolean;
  requires_supporting_document: boolean;
  recommended_document_type: string;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface LiabilityDetail extends LiabilityRecord {
  related_transactions: LiabilityRelatedTransaction[];
  related_recurring_transactions: LiabilityRelatedRecurring[];
}

export interface LiabilityListResponse {
  items: LiabilityRecord[];
  total: number;
  active_count: number;
}

export interface LiabilitySummary {
  total_assets: number;
  total_liabilities: number;
  net_worth: number;
  active_liability_count: number;
  monthly_debt_service: number;
  annual_deductible_interest: number;
}

export interface LiabilityCreatePayload {
  liability_type: LiabilityType;
  display_name: string;
  currency?: string;
  lender_name: string;
  principal_amount: number;
  outstanding_balance: number;
  interest_rate?: number;
  start_date: string;
  end_date?: string;
  monthly_payment?: number;
  tax_relevant?: boolean;
  tax_relevance_reason?: string;
  report_category?: LiabilityReportCategory;
  linked_property_id?: string;
  source_document_id?: number;
  notes?: string;
  create_recurring_plan?: boolean;
  recurring_day_of_month?: number;
}

export interface LiabilityUpdatePayload extends Partial<LiabilityCreatePayload> {
  is_active?: boolean;
}

File: backend/app/schemas/liability.py
"""Schemas for unified liability management."""
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field, field_validator

from app.models.liability import LiabilityReportCategory, LiabilitySourceType, LiabilityType


class LiabilityRelatedTransaction(BaseModel):
    id: int
    type: str
    amount: Decimal
    transaction_date: date
    description: Optional[str] = None


class LiabilityRelatedRecurring(BaseModel):
    id: int
    recurring_type: str
    description: str
    amount: Decimal
    frequency: str
    is_active: bool
    next_generation_date: Optional[date] = None


class LiabilityBase(BaseModel):
    liability_type: LiabilityType
    display_name: str = Field(..., min_length=1, max_length=255)
    currency: str = Field(default="EUR", min_length=3, max_length=3)
    lender_name: str = Field(..., min_length=1, max_length=255)
    principal_amount: Decimal = Field(..., gt=0)
    outstanding_balance: Decimal = Field(..., gt=0)
    interest_rate: Optional[Decimal] = Field(None, ge=0)
    start_date: date
    end_date: Optional[date] = None
    monthly_payment: Optional[Decimal] = Field(None, gt=0)
    tax_relevant: bool = False
    tax_relevance_reason: Optional[str] = Field(None, max_length=500)
    report_category: Optional[LiabilityReportCategory] = None
    linked_property_id: Optional[str] = None
    source_document_id: Optional[int] = None
    notes: Optional[str] = Field(None, max_length=1000)

    @field_validator("currency")
    @classmethod
    def normalize_currency(cls, value: str) -> str:
        return value.upper()


class LiabilityCreate(LiabilityBase):
    create_recurring_plan: bool = False
    recurring_day_of_month: Optional[int] = Field(None, ge=1, le=31)


class LiabilityUpdate(BaseModel):
    display_name: Optional[str] = Field(None, min_length=1, max_length=255)
    currency: Optional[str] = Field(None, min_length=3, max_length=3)
    lender_name: Optional[str] = Field(None, min_length=1, max_length=255)
    principal_amount: Optional[Decimal] = Field(None, gt=0)
    outstanding_balance: Optional[Decimal] = Field(None, ge=0)
    interest_rate: Optional[Decimal] = Field(None, ge=0)
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    monthly_payment: Optional[Decimal] = Field(None, gt=0)
    tax_relevant: Optional[bool] = None
    tax_relevance_reason: Optional[str] = Field(None, max_length=500)
    report_category: Optional[LiabilityReportCategory] = None
    linked_property_id: Optional[str] = None
    source_document_id: Optional[int] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = Field(None, max_length=1000)

    @field_validator("currency")
    @classmethod
    def normalize_currency(cls, value: Optional[str]) -> Optional[str]:
        return value.upper() if value else value


class LiabilityResponse(LiabilityBase):
    id: int
    user_id: int
    source_type: LiabilitySourceType
    report_category: LiabilityReportCategory
    linked_loan_id: Optional[int] = None
    is_active: bool
    can_edit_directly: bool
    can_deactivate_directly: bool
    edit_via_document: bool
    requires_supporting_document: bool
    recommended_document_type: str
    created_at: datetime
    updated_at: datetime

    @field_validator("linked_property_id", mode="before")
    @classmethod
    def coerce_property_id(cls, value):
        return str(value) if value is not None else None

    class Config:
        from_attributes = True


class LiabilityDetailResponse(LiabilityResponse):
    related_transactions: list[LiabilityRelatedTransaction] = []
    related_recurring_transactions: list[LiabilityRelatedRecurring] = []


class LiabilityListResponse(BaseModel):
    items: list[LiabilityResponse]
    total: int
    active_count: int


class LiabilitySummaryResponse(BaseModel):
    total_assets: Decimal
    total_liabilities: Decimal
    net_worth: Decimal
    active_liability_count: int
    monthly_debt_service: Decimal
    annual_deductible_interest: Decimal
